### PR TITLE
Add support for multiple space-separated CDATA sections in Magento\Framework\Xml\Parser

### LINKED ISSUE #29613: [Issue] Add support for multiple space-separated CDATA sections in Magento\Framework\Xml\Parser
This issue is automatically created based on existing pull request: magento/magento2#26822: Add support for multiple space-separated CDATA sections in Magento\Framework\Xml\Parser

---------

### Description (*)

Our XML files sometimes contain multiple CDATA sections per node. The XML parser class provided by Magento in [`/lib/internal/Magento/Framework/Xml/Parser.php`](https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Xml/Parser.php) already supports consecutive CDATA sections like `<![CDATA[One]]><![CDATA[Two]]><![CDATA[Three]]>`, which is correctly parsed as `OneTwoThree`. However this returns wrong results for empty text nodes in-between, like for `<![CDATA[One]]>  <![CDATA[Two]]>  <![CDATA[Three]]>`. This pull request adds the support and a related unit test.

### Related Pull Requests

The `switch` statement has been added in #681.

### Fixed Issues (if relevant)

No separate issue created.

### Manual testing scenarios (*)

Unit tests added as part of this pull request.

You can use the current 2.3.4 version for testing. The file in question [`/lib/internal/Magento/Framework/Xml/Parser.php`](https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Xml/Parser.php) is unchanged since August 2017.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by fnogatz**
### Description (*)

Our XML files sometimes contain multiple CDATA sections per node. The XML parser class provided by Magento in [`/lib/internal/Magento/Framework/Xml/Parser.php`](https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Xml/Parser.php) already supports consecutive CDATA sections like `<![CDATA[One]]><![CDATA[Two]]><![CDATA[Three]]>`, which is correctly parsed as `OneTwoThree`. However this returns wrong results for empty text nodes in-between, like for `<![CDATA[One]]>  <![CDATA[Two]]>  <![CDATA[Three]]>`. This pull request adds the support and a related unit test.

### Related Pull Requests

The `switch` statement has been added in #681.

### Fixed Issues (if relevant)

No separate issue created.

### Manual testing scenarios (*)

Unit tests added as part of this pull request.

You can use the current 2.3.4 version for testing. The file in question [`/lib/internal/Magento/Framework/Xml/Parser.php`](https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Xml/Parser.php) is unchanged since August 2017.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29613: Add support for multiple space-separated CDATA sections in Magento\Framework\Xml\Parser

### CODE REVIEW

**Comment by sidolov** [comment]
@magento create issue

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Xml/Test/Unit/_files/data.xml [hunk 1]
```diff
@@ -10,6 +10,7 @@
     <text> some text </text>
     <trim_spaces>   </trim_spaces>
     <cdata><![CDATA[  Some data here <strong>html</strong> tags are <i>allowed</i>  ]]></cdata>
+    <multiple_cdata>  <![CDATA[One]]>  <![CDATA[Two]]>  <![CDATA[Three]]>  </multiple_cdata>
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Xml/Test/Unit/_files/data.xml]: Hi @fnogatz,
Could you use more advanced test data here? Something like this:
```suggestion
    <multiple_cdata>  <![CDATA[Some data]]> here<![CDATA[ <strong>html</strong> tags]]> are <![CDATA[<i>allowed</i>]]> </multiple_cdata>
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Xml/Test/Unit/_files/data.xml (hunk 1 above)
Comment (fnogatz) [reply lib/internal/Magento/Framework/Xml/Test/Unit/_files/data.xml]: Good suggestion. It actually helped to identify another problem for the case CDATA and text nodes are mixed. Resolved via 71e0068 and 683c2e2.

**Comment by ihor-sviziev** [comment]
@magento run Unit tests, Static Tests

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @fnogatz,
Thank you for updating your PR, now it looks really good.
The only thing left - could you please also fix failing static tests? I know that this failure not caused by your changes, but still need to have not failing static tests for changed files.

Thank you!

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by fnogatz** [comment]
> The only thing left - could you please also fix failing static tests? I know that this failure not caused by your changes, but still need to have not failing static tests for changed files.

<details><summary>PHP Code Sniffer detected the following violations:</summary>

```
FOUND 14 ERRORS AFFECTING 14 LINES
----------------------------------------------------------------------
  50 | ERROR | [x] Missing short description
  51 | ERROR | [x] There must be exactly one blank line before tags
  58 | ERROR | [x] Missing short description
  59 | ERROR | [x] There must be exactly one blank line before tags
  66 | ERROR | [x] Missing short description
  67 | ERROR | [x] There must be exactly one blank line before tags
  76 | ERROR | [x] Missing short description
  77 | ERROR | [x] There must be exactly one blank line before tags
  85 | ERROR | [x] Missing short description
  86 | ERROR | [x] There must be exactly one blank line before tags
 139 | ERROR | [x] Missing short description
 140 | ERROR | [x] There must be exactly one blank line before tags
 149 | ERROR | [x] Missing short description
 150 | ERROR | [x] There must be exactly one blank line before tags
----------------------------------------------------------------------
```
</details>

The Allure Report can be found [here](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/26822/66c372ea05525e62aa73127c6cd5eb44/Statics/allure-report-b2b/index.html#suites/51d3ab4d5d77a0bc6a7ea344e9e7e6be/9e9404d5290ccd46/). Fixed via f6e09c6.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by fnogatz** [comment]
The still failing functional tests aren't related to this PR.

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests looks not related to changes form this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8171](https://jira.corp.magento.com/browse/ENGCOM-8171) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@fnogatz thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
Hi @fnogatz . During testing PR changes I tried to add multiple CDATA sections into label and there are results:
**Case 1:** add multiple CDATA into `comment` node
image
**Result:**
:x: No comment is shown under field
image

**Case 2:** add multiple CDATA into `label` node
image
**Result:**
:x: Exception error "`Notice: Array to string conversion in /var/www/html/magento2dev/lib/internal/Magento/Framework/Phrase.php on line 72` in log

Could you clarify in what cases multiple CDATA is used?

**Comment by fnogatz** [comment]
Hi @engcom-Delta. Thank you for having a look into it! We use the XML parser class provided by Magento in `/lib/internal/Magento/Framework/Xml/Parser.php` to process XML files not related to Magento, i.e. in a preprocessing step on our proprietary data-import. So the issue did not arise from a scenario where we tried to add multiple CDATA sections on Magento's files like `config.xml`, `system.xml`, `widget.xml`, or the like.

However, I tried to reproduce the issue on the Magento backend as you did by modifying Magento's XML files. To be honest, after two hours I have no clue why it behaves differently. Here's my **Case 3**:
-   Modify the `description` of `cache_lifetime` in `magento/module-catalog-widget/etc/widget.xml` to contain mixed CDATA and text, e.g., `<![CDATA[  Some data]]> here<![CDATA[ <strong>html</strong> tags]]> are <![CDATA[<i>allowed</i>  ]]>`:
    Screenshot from 2020-10-05 14-23-34
-   Even without the modifications proposed in this PR, this gets *correctly* (!) displayed in the Magento backend:
    Screenshot from 2020-10-05 14-25-13
- However, if I load this modified `widget.xml` programmatically and run `xmlToArray()`, it produces only the value `string(16) "<i>allowed</i>  "`:
    ```php
    $parser = new Parser();
    $parser->load('magento/module-catalog-widget/etc/widget.xml');
    $content = $parser->xmlToArray();    
    var_dump($content);
    /*
        ...
        ["_value"]=>
        array(2) {
          ["label"]=>
          array(2) {
            ["_value"]=>
            string(24) "Cache Lifetime (Seconds)"
            ["_attribute"]=>
            array(1) {
              ["translate"]=>
              string(4) "true"
            }
          }
          ["description"]=>
          array(2) {
            ["_value"]=>
            string(16) "<i>allowed</i>  "
            ["_attribute"]=>
            array(1) {
              ["translate"]=>
              string(4) "true"
            }
          }
        }
        ...
    */
    ```
    The expected `string(64) "  Some data here <strong>html</strong> tags are <i>allowed</i>  "` is returned only after applying this PR.

I have no idea, how the full description can be shown in the Magento backend without this PR, whereas manually loading the XML file results in an incomplete string at the same time -- which makes it hard for me to debug :/

**Comment by ihor-sviziev** [comment]
@engcom-Delta could you help with that?

**Comment by engcom-Charlie** [comment]
@fnogatz you can see merged `widget.xml` files as array for example like this
```
$widgetReader = $om->create(\Magento\Widget\Model\Config\Reader::class);
$res = $widgetReader->read();
var_dump($res);
```

**Comment by engcom-Charlie** [comment]
Hi, @fnogatz could you please provide steps to reproduce the issue?
Thank you.

**Comment by ihor-sviziev** [comment]
Hi @fnogatz,
Could you check if the issue still reproducing? Could you add steps to reproduce?

**Comment by sidolov** [comment]
@fnogatz I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**